### PR TITLE
feat: implement `get_signer_utxo` for in memory storage

### PR DESCRIPTION
**Pull request by matteojug**
## Description

Initial work for https://github.com/stacks-network/sbtc/issues/538.

I'll add the postgres impl (and more tests, if needed) in a different PR to keep this small, if there are no objections :)

## Changes

 - Move `get_signer_utxo` to storage trait (as the issue suggest to use the database)
 - Implement `get_signer_utxo` for in memory db
 - Add some tests for it

## Testing Information

Since we no longer mock it, `should_be_able_to_coordinate_signing_rounds` now also need sbtc txns to work.

## Checklist:

- [X] I have performed a self-review of my code
- [X] My changes generate no new warnings
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/storage/in_memory.rs [hunk 1]
```diff
@@ -414,6 +422,97 @@ impl super::DbRead for SharedStore {
...
+                    })
+            })
+            .next();
+
+        if sbtc_txs.is_none() {
+            return Ok(None);
+        }
+        // `sbtc_txs` contains all the txs in the highest canonical block where the first
+        // output is spendable by script_pubkey
+        let sbtc_txs = sbtc_txs.unwrap();
```
Comment (djordon) [inline signer/src/storage/in_memory.rs:477]: You could express the same thing without an unwrap, by using the let-else pattern.

File: signer/src/storage/mod.rs [hunk 2]
```diff
@@ -130,6 +131,12 @@ pub trait DbRead {
         &self,
     ) -> impl Future<Output = Result<Vec<model::Bytes>, Error>> + Send;
 
+    /// Get the outstanding signer UTXO
+    fn get_signer_utxo(
```
Comment (djordon) [inline signer/src/storage/mod.rs:135]: We should probably add a "full" description of what is fetched here.

File: signer/src/storage/mod.rs [hunk 3]
```diff
@@ -130,6 +131,12 @@ pub trait DbRead {
...
+        &self,
+        aggregate_key: &crate::keys::PublicKey,
```
Comment (djordon) [inline signer/src/storage/mod.rs:137]: This functions should take the block hash of the bitcoin chain tip as an input.

**Review by djordon — COMMENTED** [review]
Nice. Gave it a quick look over just now and will have a proper review tomorrow. Here is what I have so far.

**Review by cylewitruk — COMMENTED** [review]
Tossed in a few comments/suggestions. I didn't look too hard at the tests as I think there will be a bit of refactoring there after [this comment](https://github.com/stacks-network/sbtc/pull/567#discussion_r1774393193). Good work so far!

File: signer/src/storage/mod.rs (hunk 3 above)
Comment (matteojug) [reply signer/src/storage/mod.rs:137]: In addition to the `aggregate_key`, or only that? The issue description would suggest we "infer" (ie, pick all the valid) aggregate_key from the storage, but given it was present in the original fn signature I kept it.

Comment (djordon) [reply signer/src/storage/mod.rs:150]: Oh yeah, I meant in addition to the `aggregate_key`.

**Review by djordon — COMMENTED** [review]
Looks good.

I think the logic looks sound. But can we add another test for the case where we have a fork with chains of the same length and UTXOs on both? You can either modify an existing test or add a new one.

**Review by djordon — APPROVED** [review]
Looks good ✅

**Pull request merged**